library/Backends/llvm/src/Juvix/Backends/LLVM/Codegen/Record.hs

```diff
@@ -1 +1,132 @@
-module Juvix.Backends.LLVM.Codegen.Record where
+module Juvix.Backends.LLVM.Codegen.Record
+  ( register,
+    loadField,
+    makeRecord,
+  )
+where
+
+import qualified Data.List as List
+import qualified Juvix.Backends.LLVM.Codegen.Block as Block
+import qualified Juvix.Backends.LLVM.Codegen.Types as Types
+import qualified Juvix.Backends.LLVM.Pass.Types as PassTypes
+import Juvix.Library hiding (Type, local)
+import qualified Juvix.Library.HashMap as Map
+import Juvix.Library.NameSymbol as NameSymbol
+import qualified LLVM.AST as AST
+import qualified LLVM.AST.Type as Type
+import qualified Prelude as P
+
+-- | @register@ registers a record type with the given name and
+-- | field names and types
+register ::
+  Types.Define m =>
+  PassTypes.RecordName ->
+  [PassTypes.FieldName] ->
+  [Type.Type] ->
+  m ()
+register recordName fieldNames llvmFieldTypes = do
+  let llvmTypeName = recordTypeName recordName
+  Block.addType llvmTypeName (llvmRecordType llvmFieldTypes)
+  recordTable <- get @"recordTab"
+  let typeRef = Type.NamedTypeReference llvmTypeName
+  let fieldDescs = zip (map toSymbol fieldNames) llvmFieldTypes
+  put @"recordTab" $ Map.insert (toSymbol recordName) (typeRef, fieldDescs) recordTable
+
+recordTypeName :: NameSymbol.T -> AST.Name
+recordTypeName recordName = Block.internName $ "record-" <> toSymbol recordName
+
+llvmRecordType :: [Type.Type] -> Type.Type
+llvmRecordType llvmFieldTypes =
+  Type.StructureType
+    { isPacked = True,
+      elementTypes =
+        -- We box types, both to allow recursive structures and also to avoid
+        -- having to perform any recursive size calculations for now.
+        map Types.pointerOf llvmFieldTypes
```

Comment:
Smart choice, this will help you a lot in calculations for the sum types. Note there is a function to check for size, but this works quite well.